src/Symfony/Component/Security/Core/Authorization/Voter/SwitchRoleVoter.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\Authorization\Voter;
+
+use Symfony\Component\Security\Core\Authentication\Token\SwitchUserToken;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Authorization\AccessDecisionManager;
+
+/**
+ * SwitchRoleVoter votes if we check the impersonator roles.
+ *
+ * @author Fabien Papet <fabien.papet@gmail.com>
+ */
+class SwitchRoleVoter implements VoterInterface
+{
+    /**
+     * @var AccessDecisionManager
+     */
+    private $accessDecisionManager;
+    private $prefix;
+
+    public function __construct(AccessDecisionManager $accessDecisionManager, string $prefix = 'IMPERSONATOR_')
+    {
+        $this->accessDecisionManager = $accessDecisionManager;
+        $this->prefix = $prefix;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function vote(TokenInterface $token, $subject, array $attributes)
+    {
+        $result = VoterInterface::ACCESS_ABSTAIN;
+
+        if (!$token instanceof SwitchUserToken) {
+            return VoterInterface::ACCESS_DENIED;
+        }
+
+        foreach ($attributes as $attribute) {
+            if (!\is_string($attribute) || 0 !== strpos($attribute, $this->prefix)) {
+                continue;
+            }
+
+            return $this->accessDecisionManager->decide($token->getOriginalToken(), [substr($attribute, \strlen($this->prefix))], $subject);
```

Comment:
```php
if (VoterInterface::ACCESS_GRANTED === $result = return $this->accessDecisionManager->decide($token->getOriginalToken(), [substr($attribute, \strlen($this->prefix))], $subject)) {
    return $result;
}
```

Reply:
Ah no this should just decorate the call forget that comment.

Reply:
Hmm I think this is more complex, the current logic prevents passing two prefixed attributes, you should extend Voter to simplify the decoration or handle this case too IMHO.

Reply:
Without extending, you could go with:
```php
$attributes = array_map(function ($attribute) {
    return substr($attribute, \strlen($this->prefix));
}, array_filter($attributes, function ($attribute) {
    return 0 === strpos($attribute, $this->prefix) }));
}));

if (!$attributes) {
    return $result;
}

return $this->accessDecisionManager->decide($token->getOriginalToken(), $attributes, $subject);
```
WDYT?